src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java

```diff
@@ -0,0 +1,270 @@
+     *
+     * @param <T> The type of the target value.
+     * @param key The resource key of the target value.
+     */
+    public static <R, T extends R> DeferredHolder<R, T> create(ResourceKey<? super T> key) {
+        return new DeferredHolder<>(key);
+    }
+
+    /**
+     * Creates a new DeferredHolder with a ResourceKey.<br> Attempts to bind immediately if possible.
+     *
+     * @param key The resource key of the target object.
+     * @see #create(ResourceKey, ResourceLocation)
+     * @see #create(ResourceLocation, ResourceLocation)
+     * @see #create(ResourceKey)
+     */
+    @SuppressWarnings("unchecked")
+    protected DeferredHolder(ResourceKey<? super T> key) {
+        this.key = (ResourceKey<R>) Objects.requireNonNull(key);
+        this.bind(false);
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.<br>
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public T value() {
+        bind(true);
+        Objects.requireNonNull(this.holder, () -> "Trying to access unbound value: " + this.key);
+        return (T) this.holder.value();
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.<br>
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @Override
+    public T get() {
+        return this.value();
+    }
+
+    /**
+     * @return An optional containing the target object, if {@link #isPresent()}, otherwise {@linkplain Optional#empty() an empty optional}.
+     */
+    public Optional<T> asOptional() {
+        return isPresent() ? Optional.of(value()) : Optional.empty();
+    }
+
+    /**
+     * The type of the registry is really <? super T> but this saves us some additional ugly casting and doesn't break anything.
+     *
+     * @return The registry that this DeferredHolder is pointing at, or null if it doesn't exist.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    protected Registry<R> getRegistry() {
+        return (Registry<R>) BuiltInRegistries.REGISTRY.get(this.key.registry());
+    }
+
+    /**
+     * Binds this DeferredHolder to the underlying registry and target object.<br> Has no effect if already bound.
+     *
+     * @param throwOnMissingRegistry If true, an exception will be thrown if the registry is absent.
+     * @throws IllegalStateException If throwOnMissingRegistry is true and the backing registry is unavailable.
+     */
+    protected void bind(boolean throwOnMissingRegistry) {
+        if (this.holder != null) return;
```

Comment:
This might not be thread-safe.

Reply:
See https://github.com/neoforged/NeoForge/pull/35/files#r1335934737